### PR TITLE
implement toEthernetStreamC

### LINKED ISSUE #73: Implement IP to Ethernet stream transformer using ARP module
**Feature description**
Similar to https://github.com/GiPHouse/qbaylogic-clash-based-macipudp-stack-spring24/issues/72 but use a MAC address which is obtained from the ARP service instead of hardcoded MAC address.

**Expected behaviour**
The constructed `EthernetHeader` matches what is in the response from `ARPLookup`.

The circuit is expected to have type

```Haskell
toEthernetStream
  :: HiddenClockResetEnable dom
  => Signal dom MacAddress
  -- ^ My mac address
  -> Circuit (PacketStream dom n IPAddress) (PacketStream dom n EthernetHeader, ARPLookup dom)
```

**Expected steps**
Send out MAC Address lookups to the ARP service and use the given MAC to construct an `EthernetHeader` metatype so an `PacketStream dom n EthernetHeader` can be produced.

### PR DESCRIPTION
**Pull request by MatthijsMu**
up for review.

### CODE REVIEW

**Review by rowanG077 — APPROVED** [review]

**Comment by rowanG077** [comment]
I think not using these arcane bundles is slightly better. It means you can also flatten all these nested tuples which makes things a lot nicer to read.

```diff
diff --git a/src/Clash/Cores/Ethernet/IP/EthernetStream.hs b/src/Clash/Cores/Ethernet/IP/EthernetStream.hs
index 69cf2ba..841b924 100644
--- a/src/Clash/Cores/Ethernet/IP/EthernetStream.hs
+++ b/src/Clash/Cores/Ethernet/IP/EthernetStream.hs
@@ -7,7 +7,6 @@ import Clash.Cores.Ethernet.Arp.ArpTypes
 import Clash.Cores.Ethernet.IP.IPv4Types
 import Clash.Cores.Ethernet.Mac.EthernetTypes
 import Clash.Prelude
-import Data.Bifunctor
 import Data.Maybe ( isJust )
 import Protocols
 import Protocols.Extra.PacketStream
@@ -41,29 +40,37 @@ toEthernetStreamC myMac = fromSignals ckt
          , (Signal dom PacketStreamS2M, Signal dom  (Maybe ArpResponse)))
       -> (Signal dom PacketStreamS2M
          , (Signal dom (Maybe (PacketStreamM2S dataWidth EthernetHeader)),Signal dom (Maybe IPv4Address)))
-    ckt (packetInS, (ackInS, arpInS)) = (second unbundle . mealyB go Idle . second bundle) (myMac, packetInS, (ackInS, arpInS))
+    ckt (packetInS, (ackInS, arpInS)) = let (bwdOut, fwdOut, lookupOut) = mealyB go Idle (myMac, packetInS, ackInS, arpInS)
+                                        in (bwdOut, (fwdOut, lookupOut))
       where
         go
           :: EthernetStreamState
-          -> (MacAddress, Maybe (PacketStreamM2S dataWidth IPv4Address)
-             , (PacketStreamS2M, Maybe ArpResponse))
-          -> (EthernetStreamState, (PacketStreamS2M
-             , (Maybe (PacketStreamM2S dataWidth EthernetHeader), Maybe IPv4Address)))
-        go Idle (_, pktIn, (_, arpResponse))
+          -> ( MacAddress
+             , Maybe (PacketStreamM2S dataWidth IPv4Address)
+             , PacketStreamS2M
+             , Maybe ArpResponse
+             )
+          -> ( EthernetStreamState
+             , ( PacketStreamS2M
+               , Maybe (PacketStreamM2S dataWidth EthernetHeader)
+               , Maybe IPv4Address
+               )
+             )
+        go Idle (_, pktIn, _, arpResponse)
           = case arpResponse of
-              Nothing -> (Idle, (PacketStreamS2M False, (Nothing, fmap _meta pktIn)))
-              Just ArpEntryNotFound -> (Drop, (PacketStreamS2M False, (Nothing, Nothing)))
-              Just (ArpEntryFound ma) -> (Forward{_mac = ma}, (PacketStreamS2M False, (Nothing, Nothing)))
-        go Drop (_, pktIn, (_, _))
-          = (nextSt, (PacketStreamS2M True, (Nothing, Nothing)))
+              Nothing -> (Idle, (PacketStreamS2M False, Nothing, fmap _meta pktIn))
+              Just ArpEntryNotFound -> (Drop, (PacketStreamS2M False, Nothing, Nothing))
+              Just (ArpEntryFound ma) -> (Forward{_mac = ma}, (PacketStreamS2M False, Nothing, Nothing))
+        go Drop (_, pktIn, _, _)
+          = (nextSt, (PacketStreamS2M True, Nothing, Nothing))
           where
             pktInX = fromJustX pktIn
             nextSt =
               if isJust pktIn && isJust (_last pktInX)
               then Idle
               else Drop
-        go st@Forward{..} (mac, pktIn, (PacketStreamS2M ack, _))
-          = (nextSt, (PacketStreamS2M ack, (pktOut, Nothing)))
+        go st@Forward{..} (mac, pktIn, PacketStreamS2M ack, _)
+          = (nextSt, (PacketStreamS2M ack, pktOut, Nothing))
           where
             pktInX = fromJustX pktIn
             nextSt =
```

But this is not blocking for me. It's simply a suggestion.

**Review by Daanww — APPROVED** [review]
Looks fine

**Comment by MatthijsMu** [comment]
I have fixed the combinatorial loop mentioned in #147 and rebased on develop. I will merge as soon as CI approves.

File: src/Clash/Cores/Ethernet/IP/EthernetStream.hs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+-- outputs the packet with a completed ethernet header containing
+-- the IPv4 ether type, our IPv4 address and the looked up destination MAC.
+-- If the ARP service gave an ArpEntryNotFound, then this circuit drops the
+-- entire packet. It does not time out, instead expects the ARP service to send
+-- an ArpEntryNotFound after an appropriate timeout.
+toEthernetStreamC
+  :: forall (dom :: Domain) (dataWidth :: Nat)
+  .  HiddenClockResetEnable dom
+  => Signal dom MacAddress
+  -- ^ My Ip address
```
Comment (rowanG077) [inline src/Clash/Cores/Ethernet/IP/EthernetStream.hs]: My Mac Address

**Pull request merged**